### ISSUE TITLE
vmapping `__init__` of stateful modules

### ISSUE THREAD

**Issue by bmac3** [patrick-kidger/equinox#450]
Hi,
Thanks so much for this library! I am a huge huge fan of it

What is the recommended way to vmap over the init functions of stateful models?

As an example of what I mean, here is the way I usually initialize the modules in my models:

```
class Layer(eqx.Module):
    linear: eqx.nn.Linear
    @classmethod
    def init(cls, key):
        linear = eqx.nn.Linear(in_features=2, out_features=1, key=key)
        return cls(linear)

key = jax.random.PRNGKey(0)
three_layer_model = eqx.filter_vmap(Layer.init)(jax.random.split(key, 3))
```

This results in a model:

```
Layer(
  linear=Linear(
    weight=f32[3,1,2],
    bias=f32[3,1],
    in_features=2,
    out_features=1,
    use_bias=True
  )
)
```

This sort of initialization is especially useful when I have modules that run in parallel like multihead attention and seems to be used a lot in libraries like haliax and levanter.

Now if I make a stateful version of this layer:

```
class StatefulLayer(eqx.Module):
    linear: eqx.nn.Linear
    state_index: eqx.nn.StateIndex[Array] = eqx.field(init=False)

    def __post_init__(self):
        make_buffers = lambda **_: jnp.zeros((10, 10))
        self.state_index = eqx.nn.StateIndex(make_buffers)

    @classmethod
    def init(cls, key):
        linear = eqx.nn.Linear(in_features=2, out_features=1, key=key)
        return cls(linear)

stateful_layer = StatefulLayer.init(jax.random.PRNGKey(0))
eqx.nn.State(stateful_layer)
```

I end up with a state
```
State(0x7f7f92cb6080=f32[10,10])
```

Now if I do the same vmapped initialization with the stateful layer:

```
stateful_model = eqx.filter_vmap(StatefulLayer.init)(jax.random.split(key, 3))
eqx.nn.State(stateful_model)
```

I get out:

```
State(0x7f7f92cb6170=f32[10,10])
```

What I would expect to happen is something like

```
State(0x7f7f92cb6170=f32[3,10,10])
```
or
```
State(
  0x7f7f92cb6170=f32[10,10],
  0x7f7f92cb6171=f32[10,10],
  0x7f7f92cb6172=f32[10,10],
)
```

Then each layer would have its own state. 

Is there a recommended way to do vmapped initialization with states?

**Comment by bmac3** [comment]
I've found that initializing the state as part of my model and partitioning out the state after initialization has solved my use case so I will close this issue.

**Comment by patrick-kidger** [comment]
Actually, I'm going to reopen this one! I'm not sufficiently happy with how things are here, and you're right that this could be more ergonomic. I'm going to have a think if there's more that can be done here.

**Comment by bmac3** [comment]
Hi Patrick, great to hear that! I'm really interested to see what kind of solution you suggest for this. I actually spent a day or so thinking about how to refactor the `State` and `StateIndex` modules to fit this use case. I'll post my solution below incase you find it useful.

So the solution I came up with involves tracking all the `vmap` transformations that happens to a `StateIndex` object through a dummy variable that I called `like`:

```
class StateIndex(Module, Generic[_Value]):
    ...
    like: Array = field(init=False)

    def __init__(self, init: Callable[..., _Value]):
        ...
        self.like = jnp.empty(())
```

If I use the same example as I did above,
```
stateful_model = eqx.filter_vmap(StatefulLayer.init)(jax.random.split(key, 3))
```
will result in a model with `StateIndex`
```
StatefulLayer(
  ...
  state_index=StateIndex(
    marker=<object object at 0x7fbd30ed40e0>,
    init=<function <lambda>>,
    like=f32[3]
  )
)
```
`like` should track how the state's init function should be vmapped. `State`'s init function should then change to vmap over all of `like`'s axes:
```
class State:
    ...
    def __init__(self, model: PyTree, **kwargs):
        ...
        for leaf in leaves:
            if _is_index(leaf):
                init = partial(leaf.init, **kwargs)
                for _ in range(len(leaf.like.shape)):
                    init = jax.vmap(init)
                value = init(like=leaf.like)
```
Making these changes I was able to get a simple example to work:
```
class StatefulLayer(eqx.Module):
    linear: eqx.nn.Linear
    state_index: StateIndex[Array] = eqx.field(init=False)

    def __post_init__(self):
        make_buffers = lambda **_: jnp.ones((2,))
        self.state_index = StateIndex(make_buffers)

    def __call__(self, x, state):
        buffers = state.get(self.state_index)
        return x + buffers

    @classmethod
    def init(cls, key):
        linear = eqx.nn.Linear(in_features=2, out_features=2, key=key)
        return cls(linear)

key = jax.random.PRNGKey(0)
stateful_model = eqx.filter_vmap(StatefulLayer.init)(jax.random.split(key, 3))
state = State(stateful_model)

dynamic_model, static_model = eqx.partition(stateful_model, eqx.is_array)

def apply_layer(x, input):
    _dynamic_layer, st = input
    layer = eqx.combine(_dynamic_layer, static_model)
    return layer(x, st), None

out, _ = jax.lax.scan(apply_layer, init=jnp.ones(2), xs=(dynamic_model, state))
```
with `out` having value
```
Array([4., 4.], dtype=float32)
```
as expected.

If you like this solution I'm open to working more on it to contribute to the library (although I haven't really contributed to open source projects before so you'll have to bear with me). Anyways, thanks a ton for spending some time on this.

**Comment by patrick-kidger** [comment]
Okay, I think I've got this working now! Take a look at #466 and let me know what you think.

I started by considering a `vmap_like` approach as you suggest. (Incidentally, you can make it a bit more efficient by initialising as a size-zero array, so that you don't take up any extra memory: `jnp.empty((0,))`.) In the end I decided against this because:
- this only actually handles vmap'ing, but not any other transformation. Admittedly, right now I don't know of use-cases for transforming initialisation other than through vmap, but I wanted to be robust to anything that might arise in the future.
- with the current approach, I realised it's easy to have a subtle memory leak by doing something like
    ```python
    def __init__(...):
        some_array = jnp.array(...)
        self.index = eqx.nn.StateIndex(lambda **_: some_array)
    ```
    for which the initial `some_array` will never actually be deleted! It's forever stored in `self.index.init.__closure__`. And in fact, the built-in Equinox spectral normalisation layer did exactly this.

As such I decided that the initial state should probably be initialised at the same time as all the parameters of the model. This means changing from `StateIndex(lambda **_: initial_foo)` to just `StateIndex(initial_foo)`, and then reorganising things around that. Details in #466.

**Comment by bmac3** [comment]
Hi Patrick, this look fantastic! I agree with you that this is a cleaner solution than the `vmap_like` one. The only slight bit of awkwardness for my use case was that I had a model where you would want to instantiate a new state (or few states) every so often without having to instantiate an entire new model in memory. But I think one could simply do something like this to get around that
```
@jax.jit
def make_state(key):
    model, state = eqx.nn.make_with_state(Model)(key)
    return state
```
I believe this would only ever instantiate the state in memory, is that correct?
Anyways thanks again for the work you've done on this, I really appreciate it!

**Comment by patrick-kidger** [comment]
Yup, I think that should work!

If you get the chance to give #466 a try (before I merge it) let me know -- we can check that there's no other corner case affecting you, that I've might not have thought about.

**Comment by bmac3** [comment]
Will do! I'll try it out and let you know later this week

**Comment by bmac3** [comment]
Hi @patrick-kidger! Sorry for the late reply, last week got a bit busy for me. I tried out the new version of stateful and it works great. The only thing that didn't quite workout of my use case was the `make_state` function from my previous message. That is, when I tried running this piece of code:
`

    class Counter(eqx.Module):
      index: eqx.nn.StateIndex

      def __init__(self):
          init_state = jnp.array(0)
          self.index = eqx.nn.StateIndex(init_state)

      def __call__(self, x: Array, state: eqx.nn.State) -> tuple[Array, eqx.nn.State]:
          value = state.get(self.index)
          new_x = x + value
          new_state = state.set(self.index, value + 1)
          return new_x, new_state
  
    class Model(eqx.Module):
      linear: eqx.nn.Linear
      counter: Counter
      v_counter: Counter
  
      def __init__(self, key):
          # Not-stateful layer
          self.linear = eqx.nn.Linear(2, 2, key=key)
          # Stateful layer.
          self.counter = Counter()
          # Vmap'd stateful layer. (Whose initial state will include a batch dimension.)
          self.v_counter = eqx.filter_vmap(Counter, axis_size=2)()
  
      def __call__(self, x: Array, state: eqx.nn.State) -> tuple[Array, eqx.nn.State]:
          # This bit happens as normal.
          assert x.shape == (2,)
          x = self.linear(x)
          x, state = self.counter(x, state)
  
          # For the vmap, we have to restrict our state to just those states we want to
          # vmap, and then update the overall state again afterwards.
          #
          # After all, the state for `self.counter` isn't expecting to be batched, so we
          # have to remove that.
          substate = state.substate(self.v_counter)
          x, substate = eqx.filter_vmap(self.v_counter)(x, substate)
          state = state.update(substate)
  
          return x, state
  
    key = jr.PRNGKey(0)
    model, state = eqx.nn.make_with_state(Model)(key)
    x = jnp.array([5.0, -1.0])
    model(x, state)
    
    @jax.jit
    def make_state(key):
        _, state = eqx.nn.make_with_state(Model)(key)
        return state
    
    new_state = make_state(jr.PRNGKey(1))
    model(x, new_state)
`

It threw a key error at the last line. I think this was due to the fact that the line `state.substate(self.v_counter)` is looking for the instance of `self.v_counter` it was initialized with. Since the new state was initialized with a different instance of the model, it doesn't have the right key.

I'm not sure if there is a better way to make a new state without running into this problem, but I'll keep thinking about it for the next few days. Let me know if you have a clever solution. Thanks!

**Comment by patrick-kidger** [comment]
Hmm. The following way of remapping the keys works (and does only use public API), but it's very ugly:
```python
...
_, state = model(x, state)  # Keep old state around

def is_index(x):
    return isinstance(x, eqx.nn.StateIndex)

@eqx.filter_jit
def make_state(old_model, old_state, key):
    mapping = []

    def get_index_mapping(leaf, old_leaf):
        if type(leaf) != type(old_leaf):
            raise ValueError("New and old have different types")
        if is_index(leaf):
            mapping.append((leaf, old_leaf))

    model, state = eqx.nn.make_with_state(Model)(key)
    jax.tree_util.tree_map(get_index_mapping, model, old_model, is_leaf=is_index)
    for marker, old_marker in mapping:
        old_state = old_state.set(old_marker, state.get(marker))
    return old_state

new_state = make_state(model, state, jr.PRNGKey(1))
model(x, new_state)
```

I think the above is suffiicently hard to divine that we should find a nicer way to do this.

**Comment by patrick-kidger** [comment]
Okay, think I figured this one out! `make_with_state` just goes through and replaces the keys with integers. As each invocation of `make_with_state` iterates over its `StateIndex`s in the same order, then we get the same keys at each spot in the pytree, and things are compatible. See the latest commit on #466.

**Comment by bmac3** [comment]
Wow, great solution @patrick-kidger! I think this works perfectly and it's also quite elegant. I'm good with closing this issue now if you are. Really appreciate you working on this, helps me out a ton!

**Comment by patrick-kidger** [comment]
Great! I've just merged #466. Thank you for helping out with this, and this will appear in the upcoming v0.11.0 release of Equinox. :)
